staging/src/k8s.io/apimachinery/pkg/util/waitgroup/doc.go

```diff
@@ -0,0 +1,31 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package waitgroup re-implement sync.WaitGroup with channel
+// and does not have concurrent race issue.
+// Similarities with official sync.WaitGroup:
+// 1. Same interface [Wait, Done, Wait].
+// 2. Same Function: The main goroutine calls Add to set the number of
```

Comment:
without having looked into the detail. Why is the upstream implementation much longer? What do we loose?

Reply:
> Why is the upstream implementation much longer? 
I think is no mutex, and race pkg used 

>What do we loose?
I can not think of one. I tested and also find no issue.